README.md
# slovakia-covid19-data-py-graphs
[Version 1.0.6][changelog]

Python graphs of [open data sets](https://github.com/Institut-Zdravotnych-Analyz/covid19-data) related to the covid19 pandemic in Slovakia provided by the MoH of Slovakia.


## Graphs available
To see all graphs go to "res" folder.

1. Hospitalizations\Ventilated_Admissions_Age_WAverage_by_Vaccine_Daily.py
![Daily Slovakia Covid Hospital Ventilated Admissions - age_group Weighted Average & 7ma](./res/Hospitalizations/Ventilated_Admissions_Age_WAverage_by_Vaccine_Daily.png)

2. Hospitalizations\Ventilated_Admissions_Age_WAverage_by_Vaccine_Monthly.py
![Monthly Slovakia Covid Hospital Ventilated Admissions - age_group Weighted Average](./res/Hospitalizations/Ventilated_Admissions_Age_WAverage_by_Vaccine_Monthly.png)

3. Hospitalizations\Admissions_Age_WAverage_by_Vaccine_Daily.py
![Daily Slovakia Covid Hospital Admissions by Vaccine status and age_group Weighted Average](./res/Hospitalizations/Admissions_Age_WAverage_by_Vaccine_Daily.png)

4. Hospitalizations\Admissions_Age_WAverage_Basic_Daily.py
![Daily Slovakia Covid Hospital Admissions and age_group Weighted Average](./res/Hospitalizations/Admissions_Age_WAverage_Basic_Daily.png)

5. Hospitalizations\Admissions_Stacked_by_Age_Daily.png
![Slovakia Covid Hospital Admission Daily ma7 - Stacked by age_group](./res/Hospitalizations/Admissions_Stacked_by_Age_Daily.png)

[changelog]: ./CHANGELOG.md


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 1.0.6 - 2022-02-14

### Added
- Added python scripts for Slovakia Covid Hospital Admission Daily ma7 - Stacked by age_group.
- Added new graph figures.

### Updated
- Minor tweaks on all hospitalization scripts.
- Updated workflow daily-figures-update.yaml
- Updated graph figures.

## 1.0.5 - 2022-02-11

### Added
- Added python scripts for Daily Slovakia Covid Hospital Admissions and age_group Weighted Average.
- Added new graph figures.

### Updated
- Updated Admissions_Age_WAverage_by_Vaccine_Daily.py - Updated raw data cleansing by changing age_group to middle value, updated y limit and minor ticks on Axis 0
- Updated workflow daily-figures-update.yaml

## 1.0.4 - 2022-01-28

### Updated
- Updated workflow daily-figures-update.yaml
- Updated Admissions_Age_WAverage_by_Vaccine_Daily.py - Daily admissions axis plot changed from bars to filled area, fixed raw data by cleansing NaN

## 1.0.3 - 2022-01-26

### Added
- Added python scripts for Daily Slovakia Covid Hospital Admissions by Vaccine status and age_group Weighted Average.
- Added graph figure.

### Updated
- Updated workflow daily-figures-update.yaml

## 1.0.2 - 2022-01-22

### Added
- Added workflow daily-figures-update.yaml

### Updated
- Minor tweaks in Ventilated_Admissions_Age_WAverage_by_Vaccine_Daily.py and Ventilated_Admissions_Age_WAverage_by_Vaccine_Monthly.py

## 1.0.1 - 2022-01-19

### Updated
- Minor update of Ventilated_Admissions_Age_WAverage_by_Vaccine_Daily.py - minor plot code update, code corrections, extended date range.
- Major update of Ventilated_Admissions_Age_WAverage_by_Vaccine_Monthly.py - new plot code, added 2nd axis, extended date range.
- Changed graph figures location (prepare for automatization).

## 1.0.0 - 2021-12-09

### Added
- Added README.md file.
- Added LICENSE file.
- Added requirements.txt file.
- Added python scripts for Daily Slovakia Covid Hospital Ventilated Admissions - age_group Weighted Average & 7ma.
- Added python scripts for Monthly Slovakia Covid Hospital Ventilated Admissions - age_group Weighted Average.
- Added graph figures.

### Fixed
- UPV_Admission_Age_WAverage_by_Vaccine_Daily.py & UPV_Admission_Age_WAverage_by_Vaccine_Monthly.py:
  - Fixed raw input due source file change. [See IZA commits from 06-12-2021](https://github.com/Institut-Zdravotnych-Analyz/covid19-data/tree/main/Hospitals)

Hospitalizations/Admissions_Stacked_by_Age_Daily.py
import datetime
import requests
import array as arr
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
from matplotlib.dates import MO
import matplotlib.ticker as mticker

uri = 'https://raw.githubusercontent.com/Institut-Zdravotnych-Analyz/covid19-data/main/Hospitals/OpenData_Slovakia_Covid_Hospital_AdmissionDischarge.csv'

def is_uri_ok(uri_string, print_error=True):
  try:
    r = requests.head(uri_string)
    r.raise_for_status()
    return True
  except requests.exceptions.RequestException as error:
    if print_error:
      print(error)
    pass
  return False

if is_uri_ok(uri):
  raw = pd.read_csv(uri, sep=';', index_col=0) # error_bad_lines=False
  
  release_date = raw['Date'].values[0]
  
  # --- Set Data range -> start from date
  # row_no = raw.loc[raw['Date'] == '2021-08-01'].last_valid_index()
  # raw = raw.drop(raw.index[row_no:])

  # --- Convert date to datetime
  raw['Date'] = raw['Date'].apply(pd.to_datetime)

  # --- Cleansing raw
  raw['Vaccinated'] = raw['Vaccinated'].fillna(value='unknown') # Replace NaN with word unknown
  #raw['age_group'] = raw['age_group'].fillna(value=55) # Replace NaN with value 55 (average age of unvaccinated COVID-19 patient) - currently in use in Daily admissions axis only
  raw['age_group'] = raw['age_group']+5 # Middle value of age_group (e.g. age_group 0 = 0-10 age --> will be 5)
  raw = raw.loc[~(raw['Admissions'] == 0)] # Remove unnecessary zero admissions
  raw = raw.dropna(subset=['age_group']) # Additional cleansing before weighten average --> Remove NaN for proper calculation

  # --- Result - Axis 0
  p = pd.pivot_table(raw, index=['Date'] , columns=['age_group'], values='Admissions', aggfunc=np.sum, fill_value=0)
  p['total'] = p.sum(axis=1)

  df = p.apply(lambda x: (x / x.total)*100, axis=1)
  df = df.rolling(7, closed='left').mean()
  df = df.drop('total', axis=1)

  # --- Plot figure
  fig, ax = plt.subplots(figsize=(10, 8))
  plt.subplots_adjust(left=0.06, right=0.99, bottom= 0.065, top=0.96, hspace=0.06)

  # Common axis settings:
  ax.set_prop_cycle(color=['#EFE0FF', '#DDBEFF', '#C6DFF3', '#7EAED7', '#4F7794', '#FEE2A1', '#FDD472', '#FCB714', '#F79A7E', '#F15628', '#B5411E'])
  # Grid, major and minor ticks settings
  ax.grid(visible=True, which='both')
  ax.minorticks_on()
  ax.grid(which='major', color='#a9a9a9', linewidth=1)
  ax.grid(which='minor', color='#e0e0e0', linewidth=0.6)
  ax.tick_params(which='minor', color='#e0e0e0')
  ax.xaxis.set_major_locator(mdates.MonthLocator(interval=1))
  ax.xaxis.set_major_formatter(mdates.DateFormatter('%Y-%m'))
  ax.xaxis.set_minor_locator(mdates.WeekdayLocator(byweekday=MO))
  # Format x tick labels
  for label in ax.get_xticklabels(which='major'):
    label.set(rotation=0, horizontalalignment='center')
  # Set x axis range
  xlimoOffset = datetime.timedelta(days=7)
  ax.set_xlim(datetime.datetime(2021, 8, 1) - xlimoOffset, datetime.datetime(2022, 5, 1) + xlimoOffset)
  
  # 0. Axis - Daily Admissions Stacked by age_group
  colLabels = []
  for column in df:
    # Custom column labels
    colLabels.append(str(round(column)-5)+'-'+str(round(column)+5))
  ax.stackplot(df.index, df[5.0], df[15.0], df[25.0], df[35.0], df[45.0], df[55.0], df[65.0], df[75.0], df[85.0], df[95.0], df[105.0], labels=colLabels)
  ax.set_title('Slovakia Covid Hospital Admission Daily ma7 - Stacked by age_group', loc='center', y=1.002, x=0.5, fontsize='large')
  ax.yaxis.set_major_locator(mticker.MultipleLocator(10))
  ax.yaxis.set_minor_locator(mticker.MultipleLocator(2))
  ax.legend(loc='upper right')
  ax.set_xlabel(None)
  ax.set_ylabel("Age groups [%]")
  ax.set_ylim(0, 100)

  # Note inside plot
  ax.annotate('Source: github.com/Institut-Zdravotnych-Analyz/covid19-data (' + str(release_date) + ')',
                xy = (1.0, -0.06),
                xycoords='axes fraction',
                ha='right',
                va="center",
                fontsize=8)

  plt.savefig('./res/Hospitalizations/Admissions_Stacked_by_Age_Daily.png')

Hospitalizations/Admissions_Age_WAverage_by_Vaccine_Daily.py
import datetime
import requests
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
from matplotlib.dates import MO
import matplotlib.ticker as mticker

uri = 'https://raw.githubusercontent.com/Institut-Zdravotnych-Analyz/covid19-data/main/Hospitals/OpenData_Slovakia_Covid_Hospital_AdmissionDischarge.csv'

def is_uri_ok(uri_string, print_error=True):
  try:
    r = requests.head(uri_string)
    r.raise_for_status()
    return True
  except requests.exceptions.RequestException as error:
    if print_error:
      print(error)
    pass
  return False

if is_uri_ok(uri):
  raw = pd.read_csv(uri, sep=';', index_col=0) # error_bad_lines=False
  
  release_date = raw['Date'].values[0]
  
  # --- Set Data range -> start from date
  # row_no = raw.loc[raw['Date'] == '2021-08-01'].last_valid_index()
  # raw = raw.drop(raw.index[row_no:])

  # --- Convert date to datetime
  raw['Date'] = raw['Date'].apply(pd.to_datetime)

  # --- Cleansing raw
  raw['Vaccinated'] = raw['Vaccinated'].fillna(value='unknown') # Replace NaN with word unknown
  #raw['age_group'] = raw['age_group'].fillna(value=55) # Replace NaN with value 55 (average age of unvaccinated COVID-19 patient) - currently in use in Daily admissions axis only
  raw['age_group'] = raw['age_group']+5 # Middle value of age_group (e.g. age_group 0 = 0-10 age --> will be 5)

  # --- Sum Vaccine status (True/False) on Admissions - Axis 0
  p = pd.pivot_table(raw, index=['Date', 'age_group'] , columns=['Vaccinated'], values='Admissions', aggfunc=np.sum, fill_value=0, dropna=False)
  p = p.rename(columns={False:'unvaccinated', True:'vaccinated'})
  p = p.reset_index(1)
  p['age_group'] = p['age_group'].astype(np.int8)
  
  # - Split Data into categories
  wa_unvaccinated = p.loc[~(p['unvaccinated'] == 0)]

  wa_vaccinated = p.loc[~(p['vaccinated'] == 0)]

  unk = p.loc[~(p['unknown'] == 0)]

  # - Calculate Average by weights for each Age category
  wa_unvaccinated = wa_unvaccinated.groupby(wa_unvaccinated.index).apply(lambda x: np.average(x.age_group, weights=x.unvaccinated))
  wa_unvaccinated = wa_unvaccinated.rename('unvaccinated')

  wa_vaccinated = wa_vaccinated.groupby(wa_vaccinated.index).apply(lambda x: np.average(x.age_group, weights=x.vaccinated))
  wa_vaccinated = wa_vaccinated.rename('vaccinated')

  unk = unk.groupby(unk.index).apply(lambda x: np.average(x.age_group, weights=x.unknown))
  unk = unk.rename('unknown')

  # - Result weighten average age
  result = pd.DataFrame.join(wa_unvaccinated.to_frame(), wa_vaccinated.to_frame(), on='Date')
  result = result.join(unk.to_frame(), on='Date')

  # - Moving average
  result['unvaccinated_ma7'] = result['unvaccinated'].rolling(7, closed='left').mean()
  result['vaccinated_ma7'] = result['vaccinated'].rolling(7, closed='left').mean()
  result['unknown_ma7'] = result['unknown'].rolling(7, closed='left').mean()

  # --- Sum admissions Daily - Axis 1
  raw['age_group'] = raw['age_group'].fillna(value=55) # Cleanse raw
  pA = pd.pivot_table(raw, index=['Date', 'age_group'] , columns=['Vaccinated'], values='Admissions', aggfunc=np.sum, fill_value=0, dropna=False)
  pA = pA.rename(columns={False:'unvaccinated', True:'vaccinated'})

  resultAD = pd.DataFrame()
  resultAD['unvaccinated_adm'] = pA['unvaccinated'].groupby(level=['Date']).sum()
  resultAD['vaccinated_adm'] = pA['vaccinated'].groupby(level=['Date']).sum()
  resultAD['unknown_adm'] = pA['unknown'].groupby(level=['Date']).sum()

  # - Moving average
  resultAD['unvaccinated_adm_ma7'] = resultAD['unvaccinated_adm'].rolling(7, closed='left').mean()
  resultAD['vaccinated_adm_ma7'] = resultAD['vaccinated_adm'].rolling(7, closed='left').mean()
  resultAD['unknown_adm_ma7'] = resultAD['unknown_adm'].rolling(7, closed='left').mean()

  # print(result)
  # print(resultAD)

  # --- Plot figure
  fig, axs = plt.subplots(2, 1, figsize=(10, 8))
  plt.subplots_adjust(left=0.06, right=0.99, bottom= 0.065, top=0.99, hspace=0.06)

  # Common axis settings:
  for ax in axs:
    ax.set_prop_cycle(color=['#D9D9D9','#A4DBFD','#FFDFA4','#6b6b6b','#069af3','#ffa500'])
    # Grid, major and minor ticks settings
    ax.grid(visible=True, which='both')
    ax.minorticks_on()
    ax.grid(which='major', color='#a9a9a9', linewidth=1)
    ax.grid(which='minor', color='#e0e0e0', linewidth=0.6)
    ax.tick_params(which='minor', color='#e0e0e0')
    ax.xaxis.set_major_locator(mdates.MonthLocator(interval=1))
    ax.xaxis.set_major_formatter(mdates.DateFormatter('%Y-%m'))
    ax.xaxis.set_minor_locator(mdates.WeekdayLocator(byweekday=MO))
    # Format x tick labels
    for label in ax.get_xticklabels(which='major'):
      label.set(rotation=0, horizontalalignment='center')
    # Set x axis range
    xlimoOffset = datetime.timedelta(days=7)
    ax.set_xlim(datetime.datetime(2021, 8, 1) - xlimoOffset, datetime.datetime(2022, 5, 1) + xlimoOffset)
  
  # 0. Axis - Daily weighten average age
  ax = axs[0]
  ax.plot(result['unknown'], label='unknown')
  ax.plot(result['vaccinated'], label='vaccinated')
  ax.plot(result['unvaccinated'], label='unvaccinated')
  ax.plot(result['unknown_ma7'], label='unknown_ma7')
  ax.plot(result['vaccinated_ma7'], label='vaccinated_ma7')
  ax.plot(result['unvaccinated_ma7'], label='unvaccinated_ma7')
  ax.set_title('Slovakia Covid Hospital Admission Daily by Vaccine status - age_group Weighted Average & ma7', loc='left', y=0.9, x=0.02, fontsize='medium', backgroundcolor='white')
  ax.set_xticklabels([])
  ax.yaxis.set_major_locator(mticker.MultipleLocator(10))
  ax.yaxis.set_minor_locator(mticker.MultipleLocator(2))
  ax.legend(loc='upper right')
  ax.set_xlabel(None)
  ax.set_ylabel("Age")
  ax.set_ylim(40, 90)

  # 1. Axis - Daily admissions
  ax = axs[1]
  ax.plot(resultAD['unknown_adm'], label='unknown_adm')
  ax.plot(resultAD[['vaccinated_adm', 'unknown_adm']].sum(axis=1), label='vaccinated_adm')
  ax.plot(resultAD[['unvaccinated_adm', 'unknown_adm', 'vaccinated_adm']].sum(axis=1), label='unvaccinated_adm')
  ax.fill_between(x=resultAD['unknown_adm'].index, y1=resultAD['unknown_adm'].values)
  ax.fill_between(x=resultAD['vaccinated_adm'].index, y1=resultAD[['vaccinated_adm', 'unknown_adm']].sum(axis=1), y2=resultAD['unknown_adm'].values)
  ax.fill_between(x=resultAD['unvaccinated_adm'].index, y1=resultAD[['unvaccinated_adm', 'unknown_adm', 'vaccinated_adm']].sum(axis=1), y2=resultAD[['vaccinated_adm', 'unknown_adm']].sum(axis=1))
  ax.plot(resultAD['unknown_adm_ma7'], label='unknown_ma7')
  ax.plot(resultAD[['vaccinated_adm_ma7', 'unknown_adm_ma7']].sum(axis=1), label='vaccinated_ma7')
  ax.plot(resultAD[['unvaccinated_adm_ma7', 'unknown_adm_ma7', 'vaccinated_adm_ma7']].sum(axis=1), label='unvaccinated_ma7')
  ax.set_title('Slovakia Covid Hospital Admission Daily by Vaccine status', loc='left', y=0.9, x=0.02, fontsize='medium', backgroundcolor='white')
  ax.yaxis.set_major_locator(mticker.MultipleLocator(100))
  ax.yaxis.set_minor_locator(mticker.MultipleLocator(20))
  ax.set_xlabel(None)
  ax.set_ylabel("Admissions", labelpad=0)
  ax.set_ylim(0, 500)

  # Note inside plot
  ax.annotate('Source: github.com/Institut-Zdravotnych-Analyz/covid19-data (' + str(release_date) + ')',
                xy = (1.0, -0.12),
                xycoords='axes fraction',
                ha='right',
                va="center",
                fontsize=8)

  plt.savefig('./res/Hospitalizations/Admissions_Age_WAverage_by_Vaccine_Daily.png')

Hospitalizations/Ventilated_Admissions_Age_WAverage_by_Vaccine_Daily.py
import datetime
import requests
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import matplotlib.ticker as mticker

def is_uri_ok(uri_string, print_error=True):
  try:
    r = requests.head(uri_string)
    r.raise_for_status()
    return True
  except requests.exceptions.RequestException as error:
    if print_error:
      print(error)
    pass
  return False

uri = 'https://raw.githubusercontent.com/Institut-Zdravotnych-Analyz/covid19-data/main/Hospitals/OpenData_Slovakia_Covid_Hospital_UPV_AdmissionDischarge.csv'

if is_uri_ok(uri):
  raw = pd.read_csv(uri, sep=';', index_col=0)

  release_date = raw['Date'].values[0]

  # --- Set Data range -> start from date
  # row_no = raw.loc[raw['Date'] == '2021-08-01'].first_valid_index()
  # raw = raw.drop(raw.index[:row_no-1])

  # --- Convert date to datetime
  raw['Date'] = raw['Date'].apply(pd.to_datetime)

  # --- Create pivot table
  p = pd.pivot_table(raw, index=['Date', 'age_group'] , columns=['Vaccinated'], values='Admissions', aggfunc=np.sum, fill_value=0)
  p = p.rename(columns={False:'unvax', True:'vax'})

  # --- Split Data into categories

  # Remove rows with zero
  # When using the average function, if weights are given, they can not add up to 0 because that leads to division by 0
  wa_unvax = p.loc[~(p['unvax'] == 0)]
  # Applay reset_index to avoid MultiIndex
  wa_unvax = wa_unvax.reset_index(1)

  wa_vax = p.loc[~(p['vax'] == 0)]
  wa_vax = wa_vax.reset_index(1)

  # --- Calculate Average by weights for each category
  # https://stackoverflow.com/questions/31521027/groupby-weighted-average-and-sum-in-pandas-dataframe
  wa_unvax = wa_unvax.groupby(wa_unvax.index).apply(lambda x: np.average(x.age_group.astype(np.int8), weights=x.unvax))
  wa_unvax = wa_unvax.rename('unvaccinated') # Renamed for proper graph legend
  wa_unvax = wa_unvax.to_frame()

  wa_vax = wa_vax.groupby(wa_vax.index).apply(lambda x: np.average(x.age_group.astype(np.int8), weights=x.vax))
  wa_vax = wa_vax.rename('vaccinated')
  wa_vax = wa_vax.to_frame()

  # --- Result Data as new DataFrame
  result = pd.DataFrame.join(wa_unvax, wa_vax, on='Date')
  # print(result)
  # print(type(result))

  # --- Add moving average
  result['unvaccinated_ma7'] = result['unvaccinated'].rolling(7, closed='left').mean()
  result['vaccinated_ma7'] = result['vaccinated'].rolling(7, closed='left').mean()

  # --- Plot Data
  ax = result.plot.line(title='Slovakia Covid Hospital Ventilated Admissions - age_group Weighted Average & ma7', figsize=(10,8), color = ['#FFDFA4', '#A4DBFD', '#ffa500', '#069af3'])
  plt.subplots_adjust(left=0.06, right=0.99, bottom= 0.065, top=0.96)

  ax.grid(visible=True, which='both')
  ax.minorticks_on()

  ax.grid(which='major', color='#a9a9a9', linewidth=1)
  ax.grid(which='minor', color='#e0e0e0', linewidth=0.6)
  ax.tick_params(which='minor', color='#e0e0e0')
  ax.xaxis.set_major_locator(mdates.MonthLocator(interval=1))
  ax.xaxis.set_major_formatter(mdates.DateFormatter('%Y-%m'))
  ax.xaxis.set_minor_locator(mdates.WeekdayLocator())
  ax.yaxis.set_major_locator(mticker.MultipleLocator(10))
  ax.yaxis.set_minor_locator(mticker.MultipleLocator(2.5))
  # Format x tick labels
  for label in ax.get_xticklabels(which='major'):
    label.set(rotation=0, horizontalalignment='center')
  ax.set_xlabel(None)
  ax.set_ylabel("Age")
  # Set x and y axis range
  xlimoOffset = datetime.timedelta(days=7)
  ax.set_xlim(datetime.datetime(2021, 8, 1) - xlimoOffset, datetime.datetime(2022, 5, 1) + xlimoOffset)
  ax.set_ylim(0, 100)

  # Note inside plot area - foot note
  ax.annotate('Source: github.com/Institut-Zdravotnych-Analyz/covid19-data (' + str(release_date) + ')',
                xy = (1.0, -0.06),
                xycoords='axes fraction',
                ha='right',
                va='center',
                fontsize=8)

  plt.savefig('./res/Hospitalizations/Ventilated_Admissions_Age_WAverage_by_Vaccine_Daily.png')

Hospitalizations/Ventilated_Admissions_Age_WAverage_by_Vaccine_Monthly.py
import datetime
import requests
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import matplotlib.ticker as mticker

def is_uri_ok(uri_string, print_error=True):
  try:
    r = requests.head(uri_string)
    r.raise_for_status()
    return True
  except requests.exceptions.RequestException as error:
    if print_error:
      print(error)
    pass
  return False

uri = 'https://raw.githubusercontent.com/Institut-Zdravotnych-Analyz/covid19-data/main/Hospitals/OpenData_Slovakia_Covid_Hospital_UPV_AdmissionDischarge.csv'

if is_uri_ok(uri):
  raw = pd.read_csv(uri, sep=';', index_col=0)

  release_date = raw['Date'].values[0]

  # --- Set Data range -> start from date
  row_no = raw.loc[raw['Date'] == '2021-08-01'].last_valid_index()
  raw = raw.drop(raw.index[row_no:])

  # --- Convert date to month no.
  raw['Date'] = raw['Date'].apply(pd.to_datetime)
  raw['Date'] = raw['Date'].values.astype('datetime64[M]') # OUT: YYYY-mm-dd, result is first date in month
  # raw['Date'] = raw['Date'].to_period('M') # >>> OUT: YYYY-mm, result is not of the datetime64
  # month_grp = pd.Grouper(key='Date',freq='M') # >>> OUT: YYYY-mm-dd, result is last date in month 

  # --- Create pivot table
  p = pd.pivot_table(raw, index=['Date', 'age_group'] , columns=['Vaccinated'], values='Admissions', aggfunc=np.sum, fill_value=0)
  p = p.rename(columns={False:'unvaccinated', True:'vaccinated'})

  # --- Split Data into categories

  # Remove rows with zero
  # When using the average function, if weights are given, they can not add up to 0 because that leads to division by 0
  wa_unvax = p.loc[~(p['unvaccinated'] == 0)]
  # Applay reset_index to avoid MultiIndex
  wa_unvax = wa_unvax.reset_index(1)

  wa_vax = p.loc[~(p['vaccinated'] == 0)]
  wa_vax = wa_vax.reset_index(1)

  # --- Calculate Average by weights for each category
  # https://stackoverflow.com/questions/31521027/groupby-weighted-average-and-sum-in-pandas-dataframe
  wa_unvax = wa_unvax.groupby(wa_unvax.index).apply(lambda x: np.average(x.age_group.astype(np.int8), weights=x.unvaccinated))
  wa_unvax = wa_unvax.rename('unvaccinated') # Renamed for proper graph legend

  wa_vax = wa_vax.groupby(wa_vax.index).apply(lambda x: np.average(x.age_group.astype(np.int8), weights=x.vaccinated))
  wa_vax = wa_vax.rename('vaccinated')

  # --- Result Data as new DataFrame
  result = pd.DataFrame.join(wa_unvax.to_frame(), wa_vax.to_frame(), on='Date')
  result['unvaccinated_qty'] = p['unvaccinated'].groupby(level=['Date']).sum()
  result['vaccinated_qty'] = p['vaccinated'].groupby(level=['Date']).sum()
  # print(result.head())
  # print(type(result))

  # --- Plot
  # plt.figure("SVK C19 Hospital Ventilated Admissions")
  fig, axs = plt.subplots(2, 1, figsize=(10, 8), constrained_layout=True)

  # Custom plot layout (set constrained_layout=False)
  # plt.subplots_adjust(left=0.07, right=0.99, bottom= 0.07, top=0.99)

  # Common axis settings:
  for ax in axs:
      ax.set_prop_cycle(color=['#ffa500', '#069af3'])
      # Grid, major and minor ticks settings
      ax.grid(visible=True, which='both')
      ax.minorticks_on()
      ax.grid(which='major', color='#a9a9a9', linewidth=1)
      ax.grid(which='minor', color='#e0e0e0', linewidth=0.6)
      ax.tick_params(which='minor', color='#e0e0e0')
      ax.xaxis.set_major_locator(mdates.MonthLocator(interval=1))
      ax.xaxis.set_major_formatter(mdates.DateFormatter('%Y-%m'))
      ax.xaxis.set_minor_locator(mticker.NullLocator())
      ax.set_xlabel(None)
      # Set x axis range
      xlimoOffset = datetime.timedelta(days=7)
      ax.set_xlim(datetime.datetime(2021, 8, 1) - xlimoOffset, datetime.datetime(2022, 5, 1) + xlimoOffset)
      
  # 0. Axis - Slovakia Covid Hospital Ventilated Admissions - age_group Weighted Average
  ax = axs[0]
  ax.plot(result['unvaccinated'], label='unvaccinated', marker='o')
  ax.plot(result['vaccinated'], label='vaccinated', marker='o')
  ax.set_title('Slovakia Covid Hospital Ventilated Admissions - age_group Weighted Average', loc='left', y=0.9, x=0.02, fontsize='medium', backgroundcolor='white')
  ax.set_xticklabels([])
  ax.yaxis.set_major_locator(mticker.MultipleLocator(10))
  ax.yaxis.set_minor_locator(mticker.MultipleLocator(2.5))
  ax.legend()
  ax.set_ylabel('Age')
  ax.set_ylim(35, 85)
  # Create marker labels
  yAnnotateOffset = 0.03
  for idx, row in result.iterrows():
    ax.annotate(row['unvaccinated'].round(2), xy=(idx, row['unvaccinated']*(1-yAnnotateOffset)), fontsize=8, ha='center', va='top')
    ax.annotate(row['vaccinated'].round(2), xy=(idx, row['vaccinated']*(1+yAnnotateOffset)), fontsize=8, ha='center', va='center')

  # 1. Axis - Slovakia Covid Hospital Ventilated Admissions
  ax = axs[1]
  ax.bar(result.index, result['unvaccinated_qty'].values, bottom=result['vaccinated_qty'].values, width=10)
  ax.bar(result.index, result['vaccinated_qty'].values, width=10)
  ax.set_title('Slovakia Covid Hospital Ventilated Admissions', loc='left', y=0.9, x=0.02, fontsize='medium', backgroundcolor='white')
  ax.yaxis.set_major_locator(mticker.MultipleLocator(100))
  ax.yaxis.set_minor_locator(mticker.MultipleLocator(50))
  # Format x tick labels
  for label in ax.get_xticklabels(which='major'):
    label.set(rotation=0, horizontalalignment='center')
  ax.set_ylabel('Admissions')
  ax.set_ylim(0, 900)

  # Note inside plot area - foot note
  ax.annotate('Source: github.com/Institut-Zdravotnych-Analyz/covid19-data (' + str(release_date) + ')',
                xy = (1.0, -0.12),
                xycoords='axes fraction',
                ha='right',
                va='center',
                fontsize=8)

  plt.savefig('./res/Hospitalizations/Ventilated_Admissions_Age_WAverage_by_Vaccine_Monthly.png')

Hospitalizations/Admissions_Age_WAverage_Basic_Daily.py
import datetime
import requests
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
from matplotlib.dates import MO
import matplotlib.ticker as mticker

uri = 'https://raw.githubusercontent.com/Institut-Zdravotnych-Analyz/covid19-data/main/Hospitals/OpenData_Slovakia_Covid_Hospital_AdmissionDischarge.csv'

def is_uri_ok(uri_string, print_error=True):
  try:
    r = requests.head(uri_string)
    r.raise_for_status()
    return True
  except requests.exceptions.RequestException as error:
    if print_error:
      print(error)
    pass
  return False

if is_uri_ok(uri):
  raw = pd.read_csv(uri, sep=';', index_col=0) # error_bad_lines=False
  
  release_date = raw['Date'].values[0]
  
  # --- Set Data range -> start from date
  # row_no = raw.loc[raw['Date'] == '2021-08-01'].last_valid_index()
  # raw = raw.drop(raw.index[row_no:])

  # --- Convert date to datetime
  raw['Date'] = raw['Date'].apply(pd.to_datetime)

  # --- Cleansing raw
  raw['Vaccinated'] = raw['Vaccinated'].fillna(value='unknown') # Replace NaN with word unknown
  #raw['age_group'] = raw['age_group'].fillna(value=55) # Replace NaN with value 55 (average age of unvaccinated COVID-19 patient) - currently in use in Daily admissions axis only
  raw['age_group'] = raw['age_group']+5 # Middle value of age_group (e.g. age_group 0 = 0-10 age --> will be 5)
  raw = raw.loc[~(raw['Admissions'] == 0)] # Remove unnecessary zero admissions

  # --- Daily Admissions - Axis 1
  df = pd.DataFrame()
  df['Admissions'] = raw['Admissions'].groupby(raw['Date']).sum()
  df['Admissions_ma7'] = df['Admissions'].rolling(7, closed='left').mean()

  # --- Daily weighten average age_group - Axis 0
  raw = raw.dropna(subset=['age_group']) # Additional cleansing before weighten average --> Remove NaN for proper calculation
  df['age_group_WAverage'] = raw.groupby(raw['Date']).apply(lambda x: np.average(x.age_group, weights=x.Admissions))
  df['age_group_WAverage_ma7'] = df['age_group_WAverage'].rolling(7, closed='left').mean()


  # --- Plot figure
  fig, axs = plt.subplots(2, 1, figsize=(10, 8))
  plt.subplots_adjust(left=0.06, right=0.99, bottom= 0.065, top=0.99, hspace=0.06)

  # Common axis settings:
  for ax in axs:
    ax.set_prop_cycle(color=['#FFC1C2','#dc0002'])
    # Grid, major and minor ticks settings
    ax.grid(visible=True, which='both')
    ax.minorticks_on()
    ax.grid(which='major', color='#a9a9a9', linewidth=1)
    ax.grid(which='minor', color='#e0e0e0', linewidth=0.6)
    ax.tick_params(which='minor', color='#e0e0e0')
    ax.xaxis.set_major_locator(mdates.MonthLocator(interval=1))
    ax.xaxis.set_major_formatter(mdates.DateFormatter('%Y-%m'))
    ax.xaxis.set_minor_locator(mdates.WeekdayLocator(byweekday=MO))
    # Format x tick labels
    for label in ax.get_xticklabels(which='major'):
      label.set(rotation=0, horizontalalignment='center')
    # Set x axis range
    xlimoOffset = datetime.timedelta(days=7)
    ax.set_xlim(datetime.datetime(2021, 8, 1) - xlimoOffset, datetime.datetime(2022, 5, 1) + xlimoOffset)
  
  # 0. Axis - Daily weighten average age
  ax = axs[0]
  ax.plot(df['age_group_WAverage'], label='Daily')
  ax.plot(df['age_group_WAverage_ma7'], label='Daily_ma7')
  ax.set_title('Slovakia Covid Hospital Admission Daily - age_group Weighted Average & ma7', loc='left', y=0.9, x=0.02, fontsize='medium', backgroundcolor='white')
  ax.set_xticklabels([])
  ax.yaxis.set_major_locator(mticker.MultipleLocator(10))
  ax.yaxis.set_minor_locator(mticker.MultipleLocator(2))
  # ax.legend(loc='upper right')
  ax.set_xlabel(None)
  ax.set_ylabel("Age")
  ax.set_ylim(40, 90)

  # 1. Axis - Daily admissions
  ax = axs[1]
  ax.plot(df['Admissions'], label='Admissions')
  ax.plot(df['Admissions_ma7'], label='Admissions_ma7')
  ax.fill_between(x=df['Admissions'].index, y1=df['Admissions'].values)
  ax.set_title('Slovakia Covid Hospital Admission Daily & ma7', loc='left', y=0.9, x=0.02, fontsize='medium', backgroundcolor='white')
  ax.yaxis.set_major_locator(mticker.MultipleLocator(100))
  ax.yaxis.set_minor_locator(mticker.MultipleLocator(20))
  ax.set_xlabel(None)
  ax.set_ylabel("Admissions", labelpad=0)
  ax.set_ylim(0, 500)

  # Note inside plot
  ax.annotate('Source: github.com/Institut-Zdravotnych-Analyz/covid19-data (' + str(release_date) + ')',
                xy = (1.0, -0.12),
                xycoords='axes fraction',
                ha='right',
                va="center",
                fontsize=8)

  plt.savefig('./res/Hospitalizations/Admissions_Age_WAverage_Basic_Daily.png')